### ISSUE TITLE
Kubernetes-Vault should not take a login token via static configuration.

### ISSUE THREAD

**Issue by Timbus** [Boostport/kubernetes-vault#81]
Right now, Kube-Vault takes a login token as part of its configuration. This has some security implications as the token is prone to being exposed, for example in your deployment script.

A better approach that most of our applications use (ironically, so do pods using this controller) is they take their login details as a wrapped token. Wrapped tokens prevent leaked configuration data from being used to obtain a vault login as they are single-use and short lived.
Another approach is an operator could provide the controller with a token (wrapped or otherwise) via an API endpoint after setup.

Both approaches provide better security, but there is now a new weak point: sharing the key among replicas.

Personally I think a 'good enough' level of trust is all we can aim for. I'm not too concerned about attacks that need a lot of access already, like implementing a fake vault controller and registering it with kube. If someone does that you already have a lot of problems to worry about.
As such I think all you need is a self-signed SSL cert provided by the master, to prevent raw packet reads.. And.. maybe there could be some kind of way to verify the certificate from multiple connections, to prevent simple MITM attacks?

It seems a bit safer than the current implementation. What does everyone else think?

**Comment by F21** [comment]
In terms of providing the wrapped token to the Kubernetes-Vault controller:
- Providing the wrapped token via an API endpoint requires orchestration, which means that the onus is on the operator, or a kubernetes-vault operator would need to be developed to automate this.
- I would prefer if we renamed the `vault.token` configuration in the yaml into `vault.wrappedToken` and set the wrapped token there.

Currently, the init container generates a self-signed certificate so that the controller can send the wrapped token to it over https. I don't think there is an easy way to establish trust between the init container and the controller without operator overhead. Plus, I think perimeter security around the kubernetes cluster would be more effective in preventing the introduction of malicious pods into the cluster.

However, from the perspective of the controller trusting other instances of the controller:
- TLS or some form of encryption is a must. This can be easily achieved using self-signed certificates.
- Do we need to verify the identities of each instance of the controller before sending them the token? If we consider the kubernetes cluster to be secure and that the appropriate controls are in place to prevent malicious/fake kubernetes-vault controllers from being launched, then self-signed certificates without verification is enough. However, if verification of identities is important, then the only option is to use operator provided certificates, but this increases the amount of work the operator needs to do.
- Another problem is that it is possible for a malicious kubernetes-vault controller to join the raft cluster and get access to the token. With the previous model, this is not possible.

Regarding the modifications needed:
- In the current implementation, on start up, all instances verifies the supplied token using Vault and starts a goroutine to continually renew the token. The leader election happens independently of that.
- In order to use wrapped tokens, on start up, we wait for a leader to be elected. Once this is done, the leader unwraps the token, verifies it and starts the token renewal goroutine. It also writes the token to the raft log, which is replicated to the followers using a secure connection. The followers also verifies the token but do not start the renewal goroutine. 
- When a leader election occurs in the cluster, the leader starts the token renewal goroutine and the previous leader shuts down its token renewal goroutine.
- When a follower joins, it replicates the raft log and has access to the token.

The major decision to be made here is whether we are comfortable with using self-signed certificates for communication between kubernetes-vault instances. The risk being that the Vault token can be leaked to malicious kubernetes-vault controllers if they join the raft cluster.

**Comment by Timbus** [comment]
I think the wrappedToken value is fair. I'd go with that.

The rest of the discussion boils down to..
> Do we need to verify the identities of each instance of the controller before sending them the token?

Yes! The previous implementation relies on the fact the kubernetes cluster is secure. To keep this level of security you should verify that the kube-vault instance is a cluster member. This could be achieved by verifying that a new member has their IP listed in the service endpoint before ever allowing them to join the cluster. There's quite a few ways this could be done.. I'll have to start digging into the code on the weekend to offer sane suggestions.

**Comment by F21** [comment]
> This could be achieved by verifying that a new member has their IP listed in the service endpoint before ever allowing them to join the cluster. 

This is a good point. In fact, what we could do is make the leader a CA and have it issue certificates for the followers using the IP address as the subject.

This does lead to another chicken and egg problem though. We need a cluster to elect a leader (assuming the leader controls whether someone is allowed to join the cluster or not). However, before electing a leader, it is not possible to do this check and without a cluster, we can't have a leader.

**Comment by Timbus** [comment]
Well one automatic way to elect the leader would be to use the wrapped key... Make them race to unwrap the wrapped token. The leader would be the first controller to obtain the vault credentials. 
Seems to solve the problem nicely?

**Comment by F21** [comment]
That seems to be a nice way to do it. We'll need to check to see if we can get hashicorp/raft library to force a node to be a leader.

Would you be interested in opening a PR for this?

**Comment by Timbus** [comment]
Cool, looks like we have a spec. :D
I can certainly try to get started, but again I'll have to wait for the weekend, because I'm currently pretty busy with work. I'll make a PR after I get something going.. Or alternatively, I'll let you know if it doesn't go so well.

**Comment by Timbus** [comment]
Okay, I gave this an honest attempt on my Sunday and there are a few problems..
Ignoring the difficulties involved with moving raft/serf to TLS (it's not so straightforward). I don't think TLS is the answer. Even if the leader generates certificates using the kubernetes service IP's, it can't deliver them to the controllers without needing to make a connection.

I think the solution is to approach it exactly how the controller works with pods: Watch the IPs listed at the service endpoint, connect to new ones, and send them the cert using an out-of-band comms channel that they listen on.
And at that point, you can probably forgo using raft to share the key, and instead of sending a certificate just send them the secret token on startup.

So in short I propose this idea:
- Controllers start up and try to unwrap their secret token.
- One of them will unwrap the token, I'll deem it 'unsealed' (to fit the vault description).
- The remainders that do not get the token are 'sealed', and must open a HTTPS listener with a self-signed cert and wait for a token to be delivered.
- Unsealed controllers try to periodically connect to sealed controllers and hand them a token. They only ever try to connect to IPs listed under the kubernetes-vault service.

Not only does this solve the 'fake controller' issue, it also allows a fully sealed cluster to be unsealed manually by passing a new token through to any of them through an HTTPS connection.
(It also means I don't have to figure out the inner workings of raft/serf comms.)

Thoughts?

**Comment by F21** [comment]
Thanks for taking a stab at this one!

I think your proposal is pretty sound. I was wondering if we could try and push most of the work out to kubernetes:
- 1 controller unwraps the token and writes it to a kubernetes secret. It then creates and joins the cluster.
- The other controllers ~~polls that secret (maybe every 3 seconds)~~ watches that secret using the kubernetes API until it returns something (non-empty string). After x amount of time, if the secret is still not written, the controllers should fail.
- The other controllers verify the token against Vault and if all is good, it attempts to join the cluster. Otherwise, it exits.

I think the advantage of this is that:
- Communication using the kubernetes service account is over TLS (so we don't have to mess around with identity verification and authentication there).
- We can lean on Kubernetes secrets encrypting the token in storage (alpha in 1.7).
- Operators can use RBAC to restrict which pods have access to the secret.
- It is still possible to manually "unseal" a controller by writing the token to the secret.

The disadvantage is that we do require a newer version of kubernetes to do this, but I think trying to use the primitives available in Kubernetes is better than trying to do everything ourselves.

**Comment by Timbus** [comment]
Hm, although I can appreciate the simplicity, I can't say I want to go with that idea. We rely on Vault at our company because we can't trust kubernetes secrets. Although it's making efforts to improve the secret storage, the reality is that it's still too open for us to use. RBAC helps significantly but then the users need to ensure that it's actually setup.. I can't see it being allowed by our security team.

Basically, kubernetes-vault is our total replacement for kubernetes "secrets" -- at least until kubernetes uses Vault as its secret storage directly (they have roadmapped it). Your proposal unfortunately cancels that out, even if it is a nice idea.

**Comment by F21** [comment]
That makes sense.

With your proposal, I this the following additions would be good:
- If an instance is waiting for a token, it should have a timeout and fail once the timeout is reached.
- Only the leader should be sending tokens to "sealed" instance. This makes this easier to manage and reason about. This can be achieved by making the instance that unwrapped the token the leader.

**Comment by Timbus** [comment]
- That's a good idea.. You could also expose a health check? But really, any failure reporting would be good.
- I thought about setting a leader to dispense the token, but I wasn't sure if that was trying to be too fancy? I guess it should work out okay.

**Comment by F21** [comment]
> That's a good idea.. You could also expose a health check? But really, any failure reporting would be good.

I think just having the pod fail would be enough (exit code > 0) and keep things simple. Your monitoring should be able to pick it up and alert if something like this happens.

**Comment by swizzlr** [comment]
Just wanted to drop in my two or three cents regarding the HA/automation question:

- as a Vault operator, I have accepted that I cannot automate core operations with Vault. Same applies to Kubernetes-Vault
- Kubernetes lacks a good story for pod identity (currently, pending 1.8 I believe)
- I prefer the attack surface of "imposter K8S vault member" to "reading a configmap containing the token"

I don't know if that helps decision-making on what to do next, but you know. This project is great btw!

**Comment by F21** [comment]
Thanks for the comments @swizzlr!

- Can you link to me to the relevant issues for pod identity in 1.8?

Native kubernetes auth support recently landed in Vault. See: https://www.vaultproject.io/docs/auth/kubernetes.html

There were some discussions on twitter concerning whether the method used in Vault's implementation is dangerous or not: https://twitter.com/thockin/status/910332364567408640

I am hoping those issues can be ironed out soon, so that we won't need this project anymore 😄

**Comment by swizzlr** [comment]
@F21 same source as you, Twitter – I don't know specifically what's being worked on, I think it's the Sig Auth group. 

The new Vault K8S backend would be great if it worked with GKE, and didn't have the security concerns. For now, I'm sticking with K8S-vault – much more dependable!
